From 91ca84c862b15fe74ab9b5870e66903aec4f86dd Mon Sep 17 00:00:00 2001
From: Andrew Cooper <andrew.cooper3@citrix.com>
Date: Mon, 30 Jul 2018 14:13:09 +0200
Subject: [PATCH] x86/spec-ctrl: Fix the parsing of xpti= on fixed Intel
 hardware

The calls to xpti_init_default() in parse_xpti() are buggy.  The CPUID data
hasn't been fetched that early, and boot_cpu_has(X86_FEATURE_ARCH_CAPS) will
always evaluate false.

As a result, the default case won't disable XPTI on Intel hardware which
advertises ARCH_CAPABILITIES_RDCL_NO.

Simplify parse_xpti() to solely the setting of opt_xpti according to the
passed string, and have init_speculation_mitigations() call
xpti_init_default() if appropiate.  Drop the force parameter, and pass caps
instead, to avoid redundant re-reading of MSR_ARCH_CAPS.

Signed-off-by: Andrew Cooper <andrew.cooper3@citrix.com>
Reviewed-by: Juergen Gross <jgross@suse.com>
Reviewed-by: Wei Liu <wei.liu2@citrix.com>
Acked-by: Jan Beulich <jbeulich@suse.com>
master commit: be5e2ff6f54e0245331ed360b8786760f82fd673
master date: 2018-07-24 11:25:54 +0100
diff --git a/xen/arch/x86/spec_ctrl.c b/xen/arch/x86/spec_ctrl.c
index f59922e..dfede94 100644
--- a/xen/arch/x86/spec_ctrl.c
+++ b/xen/arch/x86/spec_ctrl.c
@@ -410,17 +410,10 @@ static bool_t __init retpoline_safe(uint64_t caps)
     }
 }
 
-static __init void xpti_init_default(bool force)
+static __init void xpti_init_default(uint64_t caps)
 {
-    uint64_t caps = 0;
-
-    if ( !force && (opt_xpti != OPT_XPTI_DEFAULT) )
-        return;
-
     if ( boot_cpu_data.x86_vendor == X86_VENDOR_AMD )
         caps = ARCH_CAPABILITIES_RDCL_NO;
-    else if ( boot_cpu_has(X86_FEATURE_ARCH_CAPS) )
-        rdmsrl(MSR_ARCH_CAPABILITIES, caps);
 
     if ( caps & ARCH_CAPABILITIES_RDCL_NO )
         opt_xpti = 0;
@@ -433,8 +426,6 @@ static __init int parse_xpti(const char *s)
     const char *ss;
     int val, rc = 0;
 
-    xpti_init_default(false);
-
     do {
         ss = strchr(s, ',');
         if ( !ss )
@@ -452,7 +443,7 @@ static __init int parse_xpti(const char *s)
 
         default:
             if ( !strcmp(s, "default") )
-                xpti_init_default(true);
+                opt_xpti = OPT_XPTI_DEFAULT;
             else if ( (val = parse_boolean("dom0", s, ss)) >= 0 )
                 opt_xpti = (opt_xpti & ~OPT_XPTI_DOM0) |
                            (val ? OPT_XPTI_DOM0 : 0);
@@ -616,7 +607,9 @@ void __init init_speculation_mitigations(void)
     if ( default_xen_spec_ctrl )
         __set_bit(X86_FEATURE_SC_MSR_IDLE, boot_cpu_data.x86_capability);
 
-    xpti_init_default(false);
+    if ( opt_xpti == OPT_XPTI_DEFAULT )
+        xpti_init_default(caps);
+
     if ( opt_xpti == 0 )
         __set_bit(X86_FEATURE_NO_XPTI, boot_cpu_data.x86_capability);
     else
